### DISCUSSION TITLE
Poor performance with a large collection and simultaneous queries

### DISCUSSION THREAD

**Discussion by simon-fauconnier** [milvus-io/milvus#17166] [Q&A and General discussion]
Hello,

We use milvus in our solution to search for similar vectors.

Our collection contains 2 fields: 
- a key (int64)
- a FloatVector of 128 dimensions (indexed in HNSW -> metric_type: IP, M: 32, efConstruction: 200)

This is our query (python) : 

```
collection.search( 
  [list of vectors (max length of 600)],
  "collection",
  param = {
	"metric_type": "IP",
	"params": {
	  "nprobe": 256,
	  "ef": 1024,
	}
  },
  limit = 5,
  expr  = None,
)
```

The milvus is deployed in cluster mode and has 2 datanodes (4cpu, 24Gi), 2 indexnodes (4 cpu, 24Gi) and 3 querynodes (8 cpu, 32Gi).
The collection contains 12,383,090 entities (but the number of elements is expected to increase significantly) and will reach for sure about 42,000,000,000 entities.

We have observed a latency that increases exponentially with the increasing number of competing requests :
- The first request is done in less than 5 seconds (looks already slow compared to some published benchmarks).
- The following requests take more and more time until they stabilize at 1m30 

We use several jobs that run in parallel and contact the milvus simultaneously
The milvus is up to date

What can we do to reduce the response time of the milvus?
Knowing that we would like to increase the number of entities and the number of jobs that perform the requests (expected to reach >300 per second)

Can the milvus hold the load with a collection of several billion entities and such query load ?

**Comment by xiaofan-luan** [comment]
> 

Hi simon, 42b seems to be a huge number and require great amount of resources~
So far our benchmark only focus on 50m and the result are published here https://milvus.io/docs/v2.0.x/benchmark.md.

To increase performance, there are a few things you may want to do:
1. wait for 2.1 release(low hanging fruit, we did a lot of optimization of the throughput  in 2.1)
2. tune your hnsw parameter, M, efconstruction and ef could be smaller for fast retrieving
3. tune segment size, the larger the faster it will be, but each segment is highly recommend not to be larger than your machine memory size * 1/8
4. partition your data. For instance, if you data can be partitioned by date and your search can be limited to single data then partition would help a lot.
5. change your SDK -> Go/Java SDK could be much faster than python in our real world test cases.
6. Use memory replica in 2.1~ It will replicate your data and then your concurrency will be doubled but again you will have to consider about cost efficienty.
7. Reduce your data, TTL might be the feature you are looking for

**Comment by majedtaki** [comment]
Hi @xiaofan-luan I have deployed milvus 2.1 (master-20220613-e9dcda16) and I specified the replica num to be 2, 4 , 10, 16 but the throughput only increased a little bit. I’m running 20 query nodes with 8c32Gi. I loaded 50M 512dim vectors and I’m using HNSWPQ which is the fastest index.

With no replicas, I can get 220 QPS. When I tried multiple replicas, the most I got was 300QPS. In my case, all the quries are nq=1.

How can I scale so milvus can do 500QPS+ with low p95 of <200ms??

**Reply by majedtaki** [reply]
Also it looks like I can get the same QPS with only 12 query nodes. So scaling out the query nodes has no effects on performance??

**Reply by xiaofan-luan** [reply]
> 

Hi majedtaki, one thing you could try is to tune concurrency. If you are using pymilvus, it's better to run multiple pymilvus instance due to the poor concurrency of python. If you use go/java, simply increase the thread pool number to see if you can get better performace

**Reply by xiaofan-luan** [reply]
If you tried to deploy the prometheus and check the CPU usage and response in servers side that might also help to investigate the issue

**Reply by xiaofan-luan** [reply]
replica set to 4 would be well enough for u case because your dataset(50M) is already large enough

**Reply by xiaofan-luan** [reply]
Also you might want to try RHNSWSQ8 which should be slightly faster than HNSWPQ with the same recall

**Reply by majedtaki** [reply]
@xiaofan-luan I have a flask server k8s deployment with 20 replica
 pods using pymilvus to do the query searches so scaling the client isn’t an issue.

I already have prometheus operator and the mivlus service monitor on. The cpu usage is about 50% for each node.  So in total it uses around 80 cpus out of 160.

In your benchmarks, with a 50m vector, what max QPS where you able to achieve with 2.1 version?

Also, does increasing the query nodes actually scale the QPS throughput or it just lowers the latency?

**Reply by majedtaki** [reply]
@xiaofan-luan Do you know why only 8 query nodes out of 20 are being utilized? The rest of the query nodes has no utilization


![Screen Shot 2022-06-17 at 11 04 48 PM](https://user-images.githubusercontent.com/81579411/174425261-7e704666-d274-4d87-af9a-3176ed78c17f.png)

**Reply by xiaofan-luan** [reply]
seems that there are still many nodes that are not really used, and rest of the nodes are not hit 800% memory bound.
you should try to 
1) increase concurrency or nq
2) use 4 replicas, each replica with 5 nodes(should be assigned automatically) If it's still not balanced we will need query coord log to understand why that's the case

**Reply by majedtaki** [reply]
@xiaofan-luan I also noticed that memory is not getting completely free after I release a collection from the query nodes. Is that normal?

**Reply by xiaofan-luan** [reply]
> @xiaofan-luan I also noticed that memory is not getting completely free after I release a collection from the query nodes. Is that normal?

It's usually hard to shrink all the memories, especially for go gc and libc alloacated memory.
But if there is a a big gap pls let us know

**Reply by majedtaki** [reply]
@xiaofan-luan I tried RHNSQW index and I only could get a throughput of 100 QPS with 20 query nodes and 4 replicas. Whats happening is that that in the query node prometheus metrics, the segcore request latency and the search segment latency are around 100ms each but in the proxy metrics, it shows the search latency to be 500ms? Also the query nodes cpu usage is very low. Its around only 100% out of 800%. If I increase the concurrency of the client, I just get higher latency and it does not increase the cpui utilization. I can't get higher than 100 QPS 

Not sure if version 2.1 has any better performance than 2.0.2 in regards of scaling throughput :(

**Reply by xiaofan-luan** [reply]
> 

what about the proxy cpu usage?

**Reply by xiaofan-luan** [reply]
> > 
> 
> what about the proxy cpu usage?

If the cpu usage is high, thinking of increasing proxy numbers

**Comment by rere950303** [comment]
@xiaofan-luan thanks for your works!
could you explain how the bigger segment improve query performance???

**Reply by xiaofan-luan** [reply]
with same a amount of data, the larger the segment, the less segment you will search.
inside each segment, the number of entity has limited impact on performance since we build index on top of data

**Reply by rere950303** [reply]
@xiaofan-luan thanks for answer!! ,, I have more some questions,,

1. I saw that a large number of collection can degrade query performance, is it true??
if so, why a large number of collection can degrade query performance?? could you explain the reason?

2. can increasing proxy replica number improve query performance>???

3. when i using partition key based search, Are only the segments in the specific partition in the boolean expression loaded into memory?

thank you in advance! 👍

**Reply by yhmo** [reply]
@rere950303 
These questions won't help you clearly understand the internal machinery of Milvus.
Some basic knowledge you need to know:

Milvus manages data by segments. By default, each segment's size is expected to be 512MB.
A segment is persisted as multiple chunks. In each chunk, a column-based file for each field.
One collection can contain multiple partitions, one partitions can contain multiple segments.
By default, each collection has a default partition.

There are two kinds of segments:
- Growing segment, can accept new data, some chunks are persisted into S3/MinIO, the last chunk is in-memory buffer to accept new data.
- Sealed segment, data is static, cannot be changed, all chunks are persisted into S3/MinIO. Small sealed segments can be merged into large segment.

A milvus cluster has these components: https://milvus.io/docs/main_components.md
Responsibility of each component:
- Pulsar/Kafka is like a WAL(write-ahead -log) component for milvus. All the inserted data will be stored in Pulsar/Kafka before consumed by datanode/querynode.
- MinIO/S3 is the major data storage of Milvus, all the original data and index data is stored here.
- Etcd stores the system metadata of Milvus, including schema of collections, runtime address of each component, paths of data/index of each segment, etc.
- Proxy receives requests from clients and dispatches requests to other components.
- Query node loads segments into memory for search, including growing segments and sealed segments. 
- Data node persists data into segments, merges small segments into large segments.
- Index node builds indexes for sealed segments.
- Root coordinator maintains data consistency and the entire system state. Milvus manages data consistency by time-tick machinery: https://milvus.io/docs/time_sync.md#Time-Synchronization. The root coordinator sends a time-tick for each 200ms, and each "data-channel" will receive the time-tick messages. Once a time-tick message is accepted by all the datanodes and querynodes, we can say the data before this time point is visible and searchable.
- Data coordinator manages data nodes, and assigns tasks to data nodes.
- Query coordinator manages query nodes and index nodes(in old versions, index coordinator manages index nodes, in new version, index coordinator is removed)


For each partition, at least one growing segment is created. When you create a collection, there is a parameter shards_num, the default value is 1, which means each partiiton has one "data-channel" to accept new data. A "data-channel" is a virtual data channel to consume data from Pulsar/Kafka to datanode or querynode.
In milvus, if a collection is loaded, all the segments will be loaded into query nodes, including growing segments and sealed segments.


When you insert some data by insert():
1. The data is received by proxy node, proxy node splits the data into pieces by shards_num(which is defined in collection schema, default is 1), and sends the pieces(as messages) into Pulsar/Kafka. The insert() returns success to the client once the data is persisted into Pulsar/Kafka.
2. For each growing segment, two "data channels" are created, one is between querynode and Pulsar/Kafka, the other is between datanode and Pulsar/Kafka.
3. Querynode and datanode subscribe messages asynchronously from Pulsar/Kafka, the data pieces are consumed by querynode and datanode.
4. For each growing segment, datanode keeps a buffer to accept new data. Once the buffer size reaches 16MB or the buffer lifetime hits the limit(10 minutes), the buffer is persisted to be a chunk into S3/MinIO.
5. If the collection is loaded, all segments are kept in query nodes. for each growing segment, querynode keeps a buffer to accept new data, so that search requests can search the data. New data is consumed by querynode from Pulsar/Kafka, only when the new data is already consumed by querynode, the data is visible&searchable by search()/query(). 
6. With more and more chunks persisted, the size of the growing segment reaches a threshold(100+MB), the growing segment is converted to be sealed segment by data coordinator. No more data is allowed to be inserted into this segment. Query node and data node will create a new buffer for the next growing segment.
7. Once a new sealed segment is generated, index node will receive a task to build index for this seal segment.
8. Once an index is finished, the query node will load this index into memory to replace the original segment data.

**Reply by yhmo** [reply]
From the above description, you can know:
With more collections created, there are more "data-channels" to be maintained. With a large number of collections, the time-tick synchronize tasks will slow down the entire system.

**Reply by rere950303** [reply]
@yhmo 
Thank you so much for your detailed explanation. So I understood that if the replica of proxy increases when there are many query requests from users, the performance will improve to some extent, is that correct? And I'm using version 2.3.x, and I saw the default shade of collection was 2, has it changed?

**Reply by yhmo** [reply]
When you call load(), there is a parameter "replica_number", default value is 1. This parameter determines how many copies of this collection are loaded in memory. If you set replica_number=2, there will be 2 copies in memory, and Milvus can balance search requests between the 2 copies. For each copy, there is a query node to lead/manage requests.


When you call search() on the client side:
1. A search request is sent to a proxy node of Milvus
2. The proxy node passes the request to a "leader query node" which is leading a copy, the query node knows the distribution of all the segments of this copy. Some segments might be loaded in other query nodes. The "leader" passes the request to other query nodes, gets the results from other query nodes, combine the results to a final result and returns the the proxy.
3. The proxy returns the final result to the client

Increase number of proxy nodes can accept more requests for each second. But the latency of request mainly depends on the performance of the query nodes. 
If you have high-performance query nodes, you can increase proxy nodes to get hight QPS. Eventually, with more and more proxy nodes, the QPS is limited by the performance of query nodes. If your query node's performance is poor, definitely it is useless to increase proxy nodes.
Increase replica_number also can increase QPS, with replica_number=2, it will requires 2X memory size to load the collection.

**Reply by yhmo** [reply]
In earlier versions of v2.3, the default value of shard_num is 2. In the later versions of v2.3, the default value has been changed to 1.

**Reply by rere950303** [reply]
@yhmo i using milvus v2.3.12 then, i understood the default value of shard_num is 1 in v2.3.12

**Comment by Yherealtita** [comment]
> Hello,
> 
> 
> 
> We use Milvus in our solution to search for similar vectors.
> 
> 
> 
> Our collection contains 2 fields: 
> 
> - a key (int64)
> 
> - a FloatVector of 128 dimensions (indexed in HNSW -> metric_type: IP, M: 32, construction: 200)
> 
> 
> 
> This is our query (python) : 
> 
> 
> 
> ```
> 
> collection.Search( 
> 
>   [list of vectors (max length of 600)],
> 
>   "collection",
> 
>   param = {
> 
> 	"metric_type": "IP",
> 
> 	"params": {
> 
> 	  "probe": 256,
> 
> 	  "ef": 1024,
> 
> 	}
> 
>   },
> 
>   limit = 5,
> 
>   expr  = None,
> 
> )
> 
> ```
> 
> 
> 
> The milvus is deployed in cluster mode and has 2 datanodes (4 CPU, 24Gi), 2 index nodes (4 CPU, 24Gi) and 3 query nodes (8 CPU, 32Gi).
> 
> The collection contains 12,383,090 entities (but the number of elements is expected to increase significantly) and will reach for sure about 42,000,000,000 entities.
> 
> 
> 
> We have observed a latency that increases exponentially with the increasing number of competing requests :
> 
> - The first request is done in less than 5 seconds (looks already slow compared to some published benchmarks).
> 
> - The following requests take more and more time until they stabilize at 1m30 
> 
> 
> 
> We use several jobs that run in parallel and contact the Milvus simultaneously
> 
> The Milvus is up to date
> 
> 
> 
> What can we do to reduce the response time of the milvus?
> 
> Knowing that we would like to increase the number of entities and the number of jobs that perform the requests (expected to reach >300 per second)
> 
> 
> 
> Can the milvus hold the load with a collection of several billion entities and such query load?

**Reply by yhmo** [reply]
No problem, Milvus can handle a collection with billions of entities and query/search.
